src/main/java/com/upserve/uppend/blobs/NativeIO.java

```diff
@@ -0,0 +1,61 @@
+package com.upserve.uppend.blobs;
+
+import jnr.ffi.*;
+import jnr.ffi.types.size_t;
+import org.slf4j.Logger;
+import com.kenai.jffi.MemoryIO;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.nio.*;
+
+public class NativeIO {
+    private static final Logger log = org.slf4j.LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private static final NativeC nativeC = LibraryLoader.create(NativeC.class).load("c");
+    static final int pageSize = nativeC.getpagesize(); // 4096 on most Linux
+
+    public enum Advice {
+        // These seem to be fairly stable https://github.com/torvalds/linux
+        // TODO add to https://github.com/jnr/jnr-constants
+        Normal(0), Random(1), Sequential(2), WillNeed(3), DontNeed(4);
+        private final int value;
+        Advice(int val) {
+            this.value = val;
+        }
+    }
+
+    public interface NativeC {
+        int madvise(@size_t long address, @size_t long size, int advice);
+        int getpagesize();
+    }
+
+    static long alignedAddress(long address) {
+        return address & (- pageSize);
+    }
+
+    static long alignedSize(long address, int capacity) {
+        long end = address + capacity;
```

Comment:
Lucene had this as `long end = alignedAddress + capacity`
This means that the pages you advise for may not cover the entire capacity of the buffer.
https://github.com/apache/lucene-solr/blob/master/lucene/misc/src/java/org/apache/lucene/store/NativePosixUtil.cpp#L305-L308

At this point I am not inclined to license this file under the lucene apache license, but open suggestions.

Reply:
I wouldn't worry about the license. IANAL, but It doesn't look like there's a single line copied verbatim.

I'm skeptical about their approach in general. It makes a lot of assumptions about alignment that don't seem necessary and could break in the future if the underlying kernel implementation changes.